### PR TITLE
Support VMService.evaluate with scope

### PR DESCRIPTION
**Pull request by annagrin**
Support VMService.evaluate  with scope by

- wrapping the expression in a function with scope keys as parameters
- compiling the wrapper
- and using chrome's Runtime.callFunctionOn to evaluate the compiled
  call with arguments

Will be followed up with a cleanup of try/catch additions and removals
in SDK and dwds.

Note that passing scope to VMService.evaluateInFrame is not supported yet.

Closes:  #1336

### CODE REVIEW

File: dwds/lib/src/debugging/inspector.dart [hunk 1]
```diff
@@ -208,6 +208,25 @@ class AppInspector extends Domain {
     return RemoteObject(result.result['result'] as Map<String, Object>);
   }
 
+  /// Calls Chrome's Runtime.callFunctionOn method with a global function.
+  ///
+  /// [evalExpression] should be a JS function definition that can accept
+  /// [arguments].
+  Future<RemoteObject> jsCallFunction(
```
Comment (grouma) [inline dwds/lib/src/debugging/inspector.dart:215]: I think this can be private.

File: dwds/lib/src/services/expression_evaluator.dart [hunk 2]
```diff
@@ -107,19 +129,24 @@ class ExpressionEvaluator {
...
-  Future<RemoteObject> evaluateExpressionInFrame(
-      String isolateId, int frameIndex, String expression) async {
+  Future<RemoteObject> evaluateExpressionInFrame(String isolateId,
+      int frameIndex, String expression, Map<String, String> scope) async {
     if (_compiler == null) {
       return _createError(ErrorKind.internal,
           'ExpressionEvaluator needs an ExpressionCompiler');
     }
 
+    if (scope != null && scope.isNotEmpty) {
```
Comment (grouma) [inline dwds/lib/src/services/expression_evaluator.dart:139]: Is this right? Why is this check necessary?

Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart:139]: I can just have a non-null check if this looks better.

Comment (grouma) [reply dwds/lib/src/services/expression_evaluator.dart:139]: Why do we support the scope for `evaluateExpression` but not in `evaluateExpressionInFrame`?

Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart:139]: As for the reason for returning error in this case - I am not sure yet how to support evaluateInFrame with scope, so I am postponing it to another task.

TL;DR:

We do support existing dart scope (the expression compiler just finds it from the file and line, no need to pass it here), but any additional overrides here will need ability to evaluate expression on a frame with object Ids as parameters. Note that chrome's `Runtime.evaluateOnCallFrame` does not support argument Ids, and `Runtime.callFunctionOn` does support argument ids but not evaluating on a different frame.

Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart]: Created an issue for that: https://github.com/dart-lang/webdev/issues/1344

File: dwds/lib/src/debugging/inspector.dart (hunk 1 above)
Comment (annagrin) [reply dwds/lib/src/debugging/inspector.dart]: Done!

File: dwds/lib/src/services/expression_evaluator.dart (hunk 2 above)
Comment (annagrin) [reply dwds/lib/src/services/expression_evaluator.dart]: `evaluate` is just lucky to have a set of matching chrome devtools API - I could wrap the expression in a function with scope as parameters and use chrome devtools `Runtime.callFunctionOn` that supports argument IDs.

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 3]
```diff
@@ -228,6 +225,10 @@ class ChromeProxyService implements VmServiceInterface {
       executionContext,
     );
 
+    _expressionEvaluator = _compiler == null
```
Comment (grouma) [inline dwds/lib/src/services/chrome_proxy_service.dart:228]: Why do we create this during an isolate instead of the chrome proxy service now?

**Review by grouma — APPROVED** [review]

File: dwds/lib/src/services/chrome_proxy_service.dart (hunk 3 above)
Comment (annagrin) [reply dwds/lib/src/services/chrome_proxy_service.dart:228]: It now uses the inspector (for `callFunction`), which is created on line 218 above, before it was only using a debugger so we could create it earlier, in the constructor for the proxy service. Postponing creating expressionEvaluator to here is fine as the clients cannot evaluate expressions before the `_initializedCompleter` is completed at  the end of this function.
